src/System.Binary.Base64/System/Binary/Base64Encoder.cs

```diff
@@ -61,10 +67,101 @@ public static OperationStatus BytesToUtf8(ReadOnlySpan<byte> bytes, Span<byte> u
+            if (symbol == 'M')  // M (for MIME) == N76
+            {
+                symbol = 'N';
+                return new ParsedFormat(symbol, MaxLineLength);
+            }
+
+            if (symbol == 'N')
+            {
+                if (format.Precision > MaxLineLength || format.Precision % 4 != 0 || format.Precision == 0)
+                {
+                    throw new FormatException($"Format {format.Symbol}:{format.Precision} not supported for Base64 Encoding.");
+                }
+                return format;
+            }
+
+            throw new FormatException($"Format {format.Symbol}:{format.Precision} not supported for Base64 Encoding.");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int BytesToUtf8Length(int bytesLength)
         {
             Debug.Assert(bytesLength >= 0);
-            return ((bytesLength + 2) / 3) << 2;
+            checked
+            {
+                return (((bytesLength + 2) / 3) * 4);
+            }
+        }
+
+        public static int BytesToUtf8Length(int bytesLength, ParsedFormat format)
+        {
+            Debug.Assert(bytesLength >= 0);
+
+            int length = BytesToUtf8Length(bytesLength);
+            if (format.IsDefault) return length;
+
+            format = ValidateFormat(format);
+
+            int bytesInOneLine = (format.Precision >> 2) * 3;
+            int extra = ((bytesLength - 1) / bytesInOneLine) * 2;
+            checked
+            {
+                return length + extra;
+            }
+        }
+
+        public static OperationStatus BytesToUtf8(ReadOnlySpan<byte> bytes, Span<byte> utf8, out int consumed, out int written, ParsedFormat format, bool isFinalBlock = true)
+        {
+            if (format.IsDefault) return BytesToUtf8(bytes, utf8, out consumed, out written, isFinalBlock);
+
+            format = ValidateFormat(format);
+
+            int inputLength = bytes.Length;
+            int lineLength = format.Precision;
+            int bytesInOneLine = (format.Precision >> 2) * 3;
+            consumed = 0;
+            written = 0;
+            OperationStatus status = OperationStatus.Done;
+
+            int numLineBreaks = utf8.Length / (lineLength + 2);
+            if ((utf8.Length % (lineLength + 2)) == (lineLength + 1))
+            {
+                numLineBreaks++;
+            }
+
+            for (int i = 0; i < numLineBreaks; i++)
+            {
+                status = BytesToUtf8(bytes.Slice(0, bytesInOneLine), utf8.Slice(0, lineLength), out int bytesConsumedLoop, out int bytesWrittenLoop);
+                utf8[lineLength] = (byte)'\r';
+                utf8[lineLength + 1] = (byte)'\n';
+                bytesWrittenLoop += 2;
+                bytes = bytes.Slice(bytesConsumedLoop);
+                utf8 = utf8.Slice(bytesWrittenLoop);
+                consumed += bytesConsumedLoop;
+                written += bytesWrittenLoop;
+            }
+
+            if (isFinalBlock)
```

Comment:
Why do we need isFinalBlock in Base64 encoding? I understand that some encodings might need it (and so our abstractions might need it), but the static method is specific  to Base64, which does not need to ever know that something is a final block.

Reply:
The last segment of Base64 data doesn't need to be 3 bytes, and padding gets added to indicate how long it was (so that the output is always a multiple of 4 characters)

Reply:
> Why do we need isFinalBlock in Base64 encoding?

The isFinalBlock is required to know when to add padding (for input length that is not a multiple of 3). Currently, we always add padding which is incorrect when partial input is passed in.